packages/ember-toucan-core/src/components/form/controls/combobox.ts

```diff
@@ -0,0 +1,537 @@
+      return;
+    }
+
+    if (
+      (event.key === 'ArrowDown' && event.metaKey) ||
+      event.key === 'PageDown' ||
+      event.key === 'End'
+    ) {
+      assert(
+        '`this.args.options` cannot be  `undefined`',
+        this.args.options !== undefined
+      );
+
+      // By default, arrowing up and down moves the insertion point to the beginning or end
+      // of an input field. We don't want this. We want to reserve arrowing for moving
+      // vertically through the list.
+      event.preventDefault();
+
+      const activeIndex = this.args.options.length - 1;
+
+      this.activeIndex = activeIndex;
+      this.#scrollActiveOptionIntoView(false);
+
+      return;
+    }
+
+    if (event.key === 'ArrowDown') {
+      assert('`this.activeIndex` cannot be `null`', this.activeIndex !== null);
+      assert(
+        '`this.args.options` cannot be  `undefined`',
+        this.args.options !== undefined
+      );
+      event.preventDefault();
+
+      const activeIndex =
+        this.activeIndex === this.args.options.length - 1
+          ? this.activeIndex
+          : this.activeIndex + 1;
+
+      this.activeIndex = activeIndex;
+
+      this.#scrollActiveOptionIntoView(false);
+
+      return;
+    }
+
+    if (
+      (event.key === 'ArrowUp' && event.metaKey) ||
+      event.key === 'PageUp' ||
+      event.key === 'Home'
+    ) {
+      event.preventDefault();
+
+      const activeIndex = 0;
+
+      this.activeIndex = activeIndex;
+      this.#scrollActiveOptionIntoView();
+
+      return;
+    }
+
+    if (event.key === 'ArrowUp') {
+      assert('`this.activeIndex` cannot be `null`', this.activeIndex !== null);
+      event.preventDefault();
+
+      const activeIndex =
+        this.activeIndex === 0 ? this.activeIndex : this.activeIndex - 1;
+
+      this.activeIndex = activeIndex;
+      this.#scrollActiveOptionIntoView();
+
+      return;
+    }
+  }
+
+  /**
+   * Handles filtering when a user types into the input element.
+   */
+  @action
+  async onInput(event: Event | InputEvent) {
+    assert(
+      'Expected HTMLInputElement',
+      event.target instanceof HTMLInputElement
+    );
+
+    const value = event.target.value;
+
+    this.inputValue = value;
+
+    const { options, optionKey, onFilter } = this.args;
+    const optionsArgument = options ? [...options] : [];
+
+    let filteredOptions: OPTION[] = [];
+
+    if (!onFilter && !optionKey) {
+      filteredOptions = (optionsArgument as string[])?.filter(
+        (option: string) =>
+          option.toLowerCase().startsWith(value.toLowerCase()) as unknown
+      ) as OPTION[];
+    }
+
+    if (!onFilter && optionKey) {
+      filteredOptions = optionsArgument?.filter((option) =>
+        ((option as Record<string, unknown>)[optionKey] as string)
+          ?.toLowerCase()
+          ?.startsWith(value.toLowerCase())
+      );
+    }
+
+    if (onFilter) {
+      filteredOptions = await onFilter(value);
```

Comment:
Should we also pass the raw options here, so the function can be simple and not need to have a of closure to have access to the options? Like

```suggestion
      filteredOptions = await onFilter(optionsArgument, value);
```
Another option (pun intended) would be to make this function a higher order function that returns the filter function, so you could do:

```suggestion
      const filterFn = onFilter(value);
      filteredOptions = optionsArgument.filter(filterFn);
```

We lose the ability for async though. But it would have the benefit that you can only return options that are actually members of `@options`. Which would be weird if that wasn't the case, and likely cause problems (e.g. when doing `===` checks). In the suggestion above such a mistake would be less likely, and more likely in the original implementation here. 

I don't have strong opinions on this though...

Reply:
Yeah for sure, I could see this being helpful.  I'm going to go ahead and merge as-is if that's okay due to our protection with changesets and am going to chat up @clintcs on Monday about this!